### PR TITLE
Version 1.2

### PR DESCRIPTION
**Pull request by clementh59**

### CODE REVIEW

File: tests/index.js [hunk 1]
```diff
@@ -354,8 +356,8 @@ describe('DigitalLink', () => {
           { rule: 'gtin-comp', match: '/01/9780345418913', remainder: '' },
           { rule: 'gtin-path', match: '/01/9780345418913', remainder: '' },
           { rule: 'gs1path', match: '/01/9780345418913', remainder: '' },
-          { rule: 'gs1uriPattern', match: '/01/9780345418913', remainder: '' },
-          { rule: 'customGS1webURI', match: 'https://gs1.evrythng.com/01/9780345418913', remainder: '' },
+          { rule: 'uncompressedGS1webURIPattern', match: '/01/9780345418913', remainder: '' },
```
[the remark this answers is not in the document]
Comment (domguinard) [reply tests/index.js:389]: The changes should be transparent for users actually but if you guys prefer we can bump to 2.0.0

File: tests/index.js [hunk 2]
```diff
@@ -472,9 +473,38 @@ describe('Utils', () => {
...
+    const input = 'https://id.gs1.org/01/47474747474747d';
     const sample = '<table class="apg-state">';
+    const res = Utils.generateResultsHtml(input);
+    expect(res).to.include(sample);
+  });
+});
 
-    expect(Utils.generateResultsHtml(input)).to.include(sample);
+describe('Grammar', () => {
+  it('should recognize https and not http', () => {
```
[the remark this answers is not in the document]
Comment (domguinard) [reply tests/index.js:515]: that's just a new test, the behavior has not changed (it covers an edge case that was present before)

[the remark this answers is not in the document]
Comment (C-D-Lewis) [reply tests/index.js:515]: Good to know!

File: tests/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (C-D-Lewis) [reply tests/index.js:389]: It is possible someone could be using `Utils.validateRule` with a now-changed rule name, but it's unlikely. Up to you.

File: src/index.js [hunk 3]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
+  //   'lot-code  = "10" / %s"lot"   ; Batch/Lot identifier',
+  //   'ser-code  = "21" / %s"ser"   ; GTIN Serial Number' ]
+  const parametersCodeRules = parameters.map(param => findTheRule(param.replace('comp', 'code')));
+
+  // I retrieve the code of each parameter : [ '22', '10', '21' ]
+  const parametersCodes = parametersCodeRules.map(singleRule => singleRule.match(/(\d+)/)[0]);
+
+  const keyQualifiersCodeWeight = parametersCodes.map((elem, index) => {
+    const str = elem.toString();
+    const map = {};
```
Comment (C-D-Lewis) [inline src/index.js:126]: This could be:
```
return {
  [str]: parameters.length - index,
};
```

File: src/index.js [hunk 4]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
+ * @returns {Map} A map that contains all the possible key qualifiers and their weights. Example : { '10': 2, '21': 1, '22': 3, cpv: 3, lot: 2, ser: 1 }
+ */
+const giveTheKeyQualifiersAWeight = () => {
+  const rule = findTheRule(Rules.gtinPath);
+
+  if (!rule) throw new Error(`The rule ${rule} wasn't find in the grammar file`);
+
+  // if my rule is : 'gtin-path  = gtin-comp  [cpv-comp] [lot-comp] [ser-comp]'
+  // In my parameters array, I'll have : [ 'cpv-comp', 'lot-comp', 'ser-comp' ]
+  const parameters = rule
```
Comment (C-D-Lewis) [inline src/index.js:108]: Would `parameterNames` be more descriptive?

File: src/index.js [hunk 5]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
 
+/**
+ * Extract from the grammar rules the order of all the possible key qualifiers and return a map with a weight for each parameter
+ *
+ * @returns {Map} A map that contains all the possible key qualifiers and their weights. Example : { '10': 2, '21': 1, '22': 3, cpv: 3, lot: 2, ser: 1 }
+ */
+const giveTheKeyQualifiersAWeight = () => {
+  const rule = findTheRule(Rules.gtinPath);
+
+  if (!rule) throw new Error(`The rule ${rule} wasn't find in the grammar file`);
```
Comment (C-D-Lewis) [inline src/index.js:104]: `found`

File: src/index.js [hunk 6]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
+
+  // if my rule is : 'gtin-path  = gtin-comp  [cpv-comp] [lot-comp] [ser-comp]'
+  // In my parameters array, I'll have : [ 'cpv-comp', 'lot-comp', 'ser-comp' ]
+  const parameters = rule
+    .substring(rule.indexOf('=') + 1)
+    .split(' ')
+    .map(item => item.replace('/ /g', ''))
```
Comment (C-D-Lewis) [inline src/index.js:111]: If this is to remove spaces, `.split(' ').join('')` could be simpler.

File: src/index.js [hunk 7]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
+    .map(item => item.replace('[', '').replace(']', ''));
+
+  // I retrieve the code rules of each parameter
+  // [ 'cpv-code  = "22" / %s"cpv"   ; Consumer Product Variant',
+  //   'lot-code  = "10" / %s"lot"   ; Batch/Lot identifier',
+  //   'ser-code  = "21" / %s"ser"   ; GTIN Serial Number' ]
+  const parametersCodeRules = parameters.map(param => findTheRule(param.replace('comp', 'code')));
+
+  // I retrieve the code of each parameter : [ '22', '10', '21' ]
+  const parametersCodes = parametersCodeRules.map(singleRule => singleRule.match(/(\d+)/)[0]);
```
Comment (C-D-Lewis) [inline src/index.js:122]: `parameterValues`?

File: src/index.js [hunk 8]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
+  const keyQualifiersCodeWeight = parametersCodes.map((elem, index) => {
+    const str = elem.toString();
+    const map = {};
+    map[str] = parameters.length - index;
+    return map;
+  });
+
+  const keyQualifiersNameWeight = parameters.map((elem, index) => {
+    const str = elem.toString().replace('-comp', '');
+    const map = {};
```
Comment (C-D-Lewis) [inline src/index.js:133]: Same strategy could be used here.

File: src/index.js [hunk 9]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
+  const keyQualifiersWeight = new Map();
+
+  // I add the two maps to the keyQualifiersWeight Map
+  keyQualifiersNameWeight.forEach(elem => {
+    Object.keys(elem).forEach(item => {
+      keyQualifiersWeight[item] = elem[item];
+    });
+  });
+  keyQualifiersCodeWeight.forEach(elem => {
+    Object.keys(elem).forEach(item => {
```
Comment (C-D-Lewis) [inline src/index.js:147]: `Object.entries(([name, value]) => { ... })` could be simpler and a bit more descriptive.

File: src/index.js [hunk 10]
```diff
@@ -65,28 +77,88 @@ const decode = (dl, str) => {
...
       });
   }
 };
 
+/**
+ * Extract from the grammar rules the order of all the possible key qualifiers and return a map with a weight for each parameter
+ *
+ * @returns {Map} A map that contains all the possible key qualifiers and their weights. Example : { '10': 2, '21': 1, '22': 3, cpv: 3, lot: 2, ser: 1 }
+ */
+const giveTheKeyQualifiersAWeight = () => {
```
Comment (C-D-Lewis) [inline src/index.js:101]: `getKeyQualifierWeights`?

File: src/util.js [hunk 11]
```diff
@@ -46,6 +52,18 @@ const validateRule = (rule, inputStr) => {
   return result.success;
 };
 
+/**
+ * The function search in the Grammar file for the rule and return it.
+ *
+ * @param {string} rule - , it has to be the name of the rule (if the line in the grammar file doesn't start with this String, the function won't find the rule)
```
Comment (C-D-Lewis) [inline src/util.js:58]: `rule - The name ...`

File: src/util.js [hunk 12]
```diff
@@ -46,6 +52,18 @@ const validateRule = (rule, inputStr) => {
...
+ * @returns {string} The rule if it is in the grammar file, otherwise undefined
+ */
+const findTheRule = rule => {
```
Comment (C-D-Lewis) [inline src/util.js:61]: `getRule (name) => `?

File: tests/index.js [hunk 13]
```diff
@@ -292,6 +292,36 @@ describe('DigitalLink', () => {
...
+      };
+
+      dl.setDomain('https://gs1.evrythng.com');
+      dl.setIdentifier(values.gtin.ai, values.gtin.value);
+      dl.setKeyQualifier(values.lot.ai, values.lot.value);
+      dl.setKeyQualifier(values.cpv.ai, values.cpv.value);
+      dl.setKeyQualifier(values.ser.ai, values.ser.value);
+
+      expect(dl.isValid()).to.equal(true);
+
```
Comment (C-D-Lewis) [inline tests/index.js:322]: Does this test really check the ordering? Or is that performed by `isValid`?

Comment (clementh59) [reply tests/index.js:322]: isValid returns false if the order is not correct.
